src/matUtils/convert.cpp

```diff
@@ -1,304 +1,267 @@
-    FILE *vcf_file = fopen(vcf_filepath.c_str(), "w");
-    std::vector<Mutation_Annotated_Tree::Node*> dfs = T.depth_first_expansion();
-    write_vcf_header(vcf_file, dfs, !no_genotypes);
-    write_vcf_rows(vcf_file, T, dfs, !no_genotypes);
-    fclose(vcf_file);
-}
-
+#include "convert.hpp"
+
+void write_vcf_header(FILE *vcf_file, std::vector<Mutation_Annotated_Tree::Node*> &dfs,
+                      bool print_genotypes) {
+    // Write minimal VCF header with sample names in same order that genotypes
+    // will be printed out (DFS).
+    fprintf(vcf_file, "##fileformat=VCFv4.2\n");
+    fprintf(vcf_file, "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO");
+    if (print_genotypes) {
+        fprintf(vcf_file, "\tFORMAT");
+        for (auto node: dfs) {
+            if (node->is_leaf()) {
+                fprintf(vcf_file, "\t%s", node->identifier.c_str());
+            }
+        }
+    }
+    fputc('\n', vcf_file);
+}
+
+uint count_leaves(std::vector<Mutation_Annotated_Tree::Node*> &dfs) {
+    // Return the number of leaf nodes in dfs
+    uint count = 0;
+    for (auto node: dfs) {
+        if (node->is_leaf()) {
+            count++;
+        }
+    }
+    return count;
+}
+
+int8_t *new_gt_array(int size, int8_t ref) {
+    // Allocate and return an array of int8_t (encoded nucleotide values) initialized to ref.
+    int8_t *gt_array = new int8_t[size];
+    for (int i = 0;  i < size;  i++) {
+        gt_array[i] = ref;
+    }
+    return gt_array;
+}
+
+uint r_add_genotypes(MAT::Node *node,
+                     std::unordered_map<std::string, std::vector<int8_t *>> &chrom_pos_genotypes, 
+                     std::unordered_map<std::string, std::vector<int8_t>> &chrom_pos_ref,
+                     uint leaf_count, uint leaf_ix, std::vector<struct MAT::Mutation *> &mut_stack) {
+    // Traverse tree, adding leaf/sample genotypes for mutations annotated on path from root to node
+    // to chrom_pos_genotypes (and reference allele to chrom_pos_ref).
+    for (auto &mut: node->mutations) {
+      if (mut.is_masked()) {
+          continue;
+      }
+      mut_stack.push_back(&mut);
+    }
+    if (node->is_leaf()) {
+        // Store genotypes in this leaf's column for all mutations on the path from root to leaf
+        for (auto mut: mut_stack) {
+            std::string chrom = mut->chrom;
+            uint pos = (uint)mut->position;
+            if (chrom.empty()) {
+              fprintf(stderr, "mut->chrom is empty std::string at node '%s', position %u\n",
+                      node->identifier.c_str(), pos);
+            }
+            if (chrom_pos_genotypes.find(chrom) == chrom_pos_genotypes.end()) {
+                // First variant on chrom: initialize a vector mapping position to genotype.
+                // Assume a genome size similar to SARS-CoV-2, resize if necessary.
+                uint initSize = 30000;

```

Comment:
You should not use hard-coded numbers like this. Take the genome size as a REQUIRED parameter from the user if you need it.

Reply:
Angie wrote the VCF conversion code and I haven't delved into it. Thoughts @AngieHinrichs ?

Reply:
30000 is just the initial size; if it turns out to be not large enough, the arrays are resized to twice their current size (a few lines below the quoted lines).  So it won't fail with larger genomes, although it would go through ceil(log2(genomeSize/30000)) rounds of resizing.

So how about an optional --genome-size param?

Reply:
Ah, not a big problem then as long it is not going to fail.